# 03 – Feature Engineering

### Purpose
This notebook creates engineered features that enhance the predictive power and interpretability of the TED CAN dataset.
All features are based on insights from Notebook 01 (EDA) and domain knowledge of public procurement.

### Steps
- Load cleaned‑saved datasets
- Apply feature engineering pypline
- Save full datasets

--------------------
#### Imports & Setup & Dataset
-------------------

In [1]:
# ---------------------------------------------------------
# Import moduls
# ---------------------------------------------------------


# import standard modules
import pandas as pd
import numpy as np
from pathlib import Path

import sys
from pathlib import Path

In [2]:
# shut off some annoying warnings
import warnings

warnings.filterwarnings("ignore", message="A value is trying to be set on a copy")
warnings.filterwarnings("ignore", category=UserWarning)
warnings.filterwarnings("ignore", category=FutureWarning)

In [3]:
# ---------------------------------------------------------
# Setup style
# ---------------------------------------------------------

# show all columns
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# visualisation settings
pd.set_option('display.float_format', '{:,.2f}'.format)

In [4]:
# ---------------------------------------------------------
# Load scripts
# ----------------------------------------------------------

%load_ext autoreload
%autoreload 2

from pathlib import Path
import sys

# found min directory
PROJECT_ROOT = Path("..").resolve()

# maindirectory sys.path
if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))

# import function from script
from my_scripts.feature_engineering import feature_engineering
from my_scripts.eda import overview

In [5]:
# ---------------------------------------------------------
# Import dataset
# ---------------------------------------------------------

df = pd.read_pickle("../data/dataset_clean.pkl")

print("EU dataset:", df.shape)

EU dataset: (3003042, 12)


---------------------------------------------------------
### Apply feature engineering

-----------------------------

In [6]:
# apply funktion
df_fe = feature_engineering(df)


In [7]:
# shape of the dataset
print(df_fe.shape)

(3003042, 13)


In [8]:
# inspekt of the EU dataset
overview(df_fe)

,dtype,total,missing_n,missing_%,uniques_n,uniques
YEAR,int16,3003042,0,0.00,9,"[2008, 2009, 2010, 2011, 2012, 2013, 2014, 201..."
ISO_COUNTRY_CODE,str,3003042,0,0.00,33,"[DE, ES, FR, PL, HU, IT, CY, UK, RO, PT, SE, N..."
TYPE_OF_CONTRACT,str,3003042,0,0.00,3,"[W, U, S]"
TOP_TYPE,str,3003042,0,0.00,10,"[OPE, NIC, RES, NOC, Unknown, COD, AWP, NOP, N..."
AWARD_VALUE_EURO,float64,3003042,0,0.00,1603450,"[302964.15, 478780.08, 14237368.1, 5204457.56,..."
CRIT_PRICE_WEIGHT,int64,3003042,0,0.00,2,"[1, 0]"
NUMBER_OFFERS,float64,3003042,0,0.00,363,"[2.0, 3.0, 6.0, 9.0, 45.0, 1.0, 5.0, 4.0, 10.0..."
TITLE,str,3003042,0,0.00,970607,"[Unknown, Pakiet Nr 4., Pakiet Nr 5., Pakiet N..."
IS_FAILED_TENDER,int64,3003042,0,0.00,2,"[0, 1]"
AWARD_QUARTER,float64,3003042,0,0.00,4,"[3.0, 4.0, 1.0, 2.0]"


#### Notes: Feature Engineering Summary
Feature engineering enriches the cleaned dataset with additional analytical variables that improve modelling of tender competition, risk, and procedural behaviour.

1. Dataset Size After Features Engineering
  - Before: 3,003,042 rows × 12 columns
  - After: 3,003,042 rows × 14 columns

2. Created Features
- Date‑based features
  - AWARD_QUARTER
  - DAYS_TO_AWARD (time between dispatch and award)
These variables capture temporal patterns, seasonality, and procedural duration.

- CPV map with 12 group
  - "Construction & Building Works": 45–46
  - "Machinery & Industrial Equipment": 30–32
  - "Electrical, Optical & Precision Equipment": 33–34
  - "IT, Software & Telecommunications": 48–50
  - "Transport Services & Logistics": 60–63
  - "Business, Consulting & Administrative Services": 79–80
  - "Cleaning, Facility & Maintenance Services": 90–91
  - "Energy, Utilities & Environmental Services": 92–99
  - "Other"

- CAE 
  - "1": "Ministry",
  - "3": "National Agency",
  - "4": "Regional/Local Authority",
  - "5": "Regional/Local Agency",
  - "5A": "Local Agency",
  - "6": "Public Body",
  - "8": "Public Undertaking",
  - "N": "Not Specified",
  - "R": "Other Public Authority",
  - "Z": "Unknown"

- Missing‑information flags
  - CRIT_PRICE_WEIGHT_MISSING
These indicators allow the model to capture informative missingness.

3. Missing numiric values replaced by the median.

4. Removed Raw Columns
After generating engineered features, the following original fields were dropped to avoid redundancy:
  - "DT_AWARD", "DT_DISPATCH"
  - "MAIN_CPV_CODE_GPA"
  - "CAE_TYPE"


------------------
### Save engineered datasets

-------------------

In [9]:
df_fe.to_pickle("../data/dataset_fe.pkl")